kernel/mmanutils.c

```diff
@@ -862,3 +884,160 @@ void myst_mman_stats(myst_mman_stats_t* buf)
+        ERAISE(-EINVAL);
+
+    /* addr must be aligned on a page boundary */
+    if ((uint64_t)addr % PAGE_SIZE)
+        ERAISE(-EINVAL);
+
+    /* round length up to the next multiple of the page boundary */
+    ECHECK(myst_round_up(length, PAGE_SIZE, &length));
+
+    myst_spin_lock(&_mman.lock);
+    locked = true;
+
+    // Fail if [addr:length] is not within the mmap area.
+    // Calculate the index of pids[] and count to set.
+    {
+        const myst_kernel_args_t* kargs = &__myst_kernel_args;
+        const uint64_t addr_start = (uint64_t)addr;
+        const uint64_t addr_end;
+        const uint64_t mman_start = (uint64_t)kargs->mman_data;
+        const size_t mman_end;
+
+        if (__builtin_add_overflow(addr_start, length, &addr_end))
+            ERAISE(-ERANGE);
+
+        if (__builtin_add_overflow(mman_start, kargs->mman_size, &mman_end))
+            ERAISE(-ERANGE);
+
+        if (!(addr_start >= mman_start && addr_end <= mman_end))
+            ERAISE(-EINVAL);
+
+        index = (addr_start - mman_start) / PAGE_SIZE;
+        count = length / PAGE_SIZE;
+
+        if (index >= npids)
+            ERAISE(-ERANGE);
+
+        if (index + count > npids)
+            ERAISE(-ERANGE);
+    }
+
+    switch (op)
+    {
+        case MMAN_PIDS_OP_SET:
+        {
+            /* Update the associated elements of pids[] */
+            for (size_t i = index; i < index + count; i++)
+            {
+                pids[i] = pid;
+            }
+
+            break;
+        }
+        case MMAN_PIDS_OP_TEST:
+        {
+            /* Update the associated elements of pids[] */
+            for (size_t i = index; i < index + count; i++)
+            {
+                if (pids[i] != (uint32_t)pid)
+                    ERAISE(-ENOENT);
+            }
+
+            break;
+        }
+        default:
+        {
+            ERAISE(-EINVAL);
+            break;
+        }
+    }
+
+done:
+
+    if (locked)
+        myst_spin_unlock(&_mman.lock);
+
+    // myst_set_trace(false);
+    return ret;
+}
+
+int myst_mman_pids_set(const void* addr, size_t length, pid_t pid)
+{
+    return _handle_mman_pids_op(MMAN_PIDS_OP_SET, addr, length, pid);
+}
+
+int myst_mman_pids_test(const void* addr, size_t length, pid_t pid)
+{
+    return _handle_mman_pids_op(MMAN_PIDS_OP_TEST, addr, length, pid);
+}
+
+int myst_mman_pids_munmap(const void* addr, size_t length, pid_t pid)
+{
+    int ret = 0;
+
+    if (!addr || pid < 0)
+        ERAISE(-EINVAL);
+
+    /* allowing a zero-valued pid would remove kernel mappings */
+    if (pid == 0)
+        ERAISE(-EINVAL);
+
+    /* addr must be aligned on a page boundary */
+    if ((uint64_t)addr % PAGE_SIZE)
+        ERAISE(-EINVAL);
+
+    /* the length must be a multiple of the page boundary */
+    if (length % PAGE_SIZE)
+        ERAISE(-EINVAL);
+
+    /* release every page in the mapping owned by this pid */
+    /* ATTN: optimize to handle consecutive pages by the same owner */
```

Comment:
Need this optimization. unmap() always triggers mprotect() OCALL. Calling unmap() page-by-page will see performance degradation.

Reply:
Fixed, thanks.